### PR TITLE
Proposal: Improve branch selection

### PR DESCRIPTION
**Pull request by alejandrohdezma**
# What has been done in this PR?

This PR improves the branch selection feature added in #2183 by allowing to be used not only for users of the [`scala-steward-action`](https://github.com/scala-steward-org/scala-steward-action) or users that run their own instance of Steward, but to anyone using the [`repos.md`](https://github.com/scala-steward-org/repos/blob/main/repos-github.md) feature.

# How?

Instead of receiving the default branch to use from the CLI, I've updated the `repos.md` parser to also allow the following syntax:

```markdown
- organization/repository:branch
```

The `Repo` class has been updated with a new field `branch: Option[Branch]`. If that field is present, that branch will be used as the default branch instead of actual default branch.

# What does this change mean?

This change means that repositories like cats or http4s (or any repository using multibranch) can maintain several branches updated with Scala Steward by updating the `repos.md` line to:

```diff
- http4s/http4s
+ http4s/http4s:main
+ http4s/http4s:0.23.x
+ http4s/http4s:0.22.x
+ http4s/http4s:0.21.x
```

## Isn't this a breaking change?

Removing the CLI param doesn't provoke a breaking change, since a new version hasn't been released since #2183 was merged 😸.

# Want to see it in action?

[Here](https://github.com/alejandrohdezma/scala-steward-new-feature-test/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc) you can see this new feature in action. Steward was run with the following parameters:

```
--do-not-fork --workspace workspace --repos-file repos.md --git-author-email me@scala-steward.org --vcs-login scala-steward --git-ask-pass askpass.sh
```

with `repos.md` being:

```md
- alejandrohdezma/scala-steward-new-feature-test
- alejandrohdezma/scala-steward-new-feature-test:0.21.x
- alejandrohdezma/scala-steward-new-feature-test:0.22.x
```

As you can see [here](https://github.com/alejandrohdezma/scala-steward-new-feature-test/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc) Scala Steward created 3 PRs, each pointing to a different branch, and each one of them using the `.scala-steward.conf` present in that branch.

### CODE REVIEW

**Comment by fthomas** [comment]
I really would like to have this feature in Scala Steward but more changes are needed to properly support multiple branches. (Some of them were in #1469.) For example, for each branch we need to have different `RepoCache`s that are persisted to disk. I think in this PR, the `RepoCache`s for each branch is persisted to the same file and thus overwriting the cache of a prior branch.

**Comment by alejandrohdezma** [comment]
Hey @fthomas I've already included all the missing changes in #1469. I believe the cache issue is resolved, given that the cache is created using the `Repo` as key, and the `Repo` contains the custom branch (if any is present), but if I'm misinterpreting something, let me know.

The only missing thing would be to try this using fork (I've already tried it using the `doNotFork`). Do you have any kind of playground for testing this kind of changes? If not, I can set something up quickly.

Once we are happy with the solution, I can update README and documentation, and create a PR for the GitHub Action.

**Comment by alejandrohdezma** [comment]
Also, it seems the [`KeyDecoder[Repo]`](https://github.com/scala-steward-org/scala-steward/blob/49fe121313bca7dd74a2d48061c77d18c09567f7/modules/core/src/main/scala/org/scalasteward/core/vcs/data/Repo.scala#L33) is unused. Should I remove it? Or is it there for some reason?

**Comment by fthomas** [comment]
> Also, it seems the [`KeyDecoder[Repo]`](https://github.com/scala-steward-org/scala-steward/blob/49fe121313bca7dd74a2d48061c77d18c09567f7/modules/core/src/main/scala/org/scalasteward/core/vcs/data/Repo.scala#L33) is unused. Should I remove it? Or is it there for some reason?

It was added in this ancient [commit](https://github.com/scala-steward-org/scala-steward/commit/f2159950b5d06d69dc1f98503a4fac6c2a4afb92) and was required for the `Decoder[Store]` instance there. It was then later required for the `JsonKeyValueStore` which lost that requirement in https://github.com/scala-steward-org/scala-steward/pull/1220. It looks indeed unused now and can be removed.

**Comment by fthomas** [comment]
I just tried it locally:
```
[info] 2021-10-16 16:15:11,582 INFO  ──────────── Steward scala-steward-org/test-repo-2/test-branch ────────────
[info] 2021-10-16 16:15:11,583 INFO  Check cache of scala-steward-org/test-repo-2/test-branch
[info] 2021-10-16 16:15:12,184 INFO  Clone scala-steward/test-repo-2
[info] 2021-10-16 16:15:12,876 ERROR Steward scala-steward-org/test-repo-2/test-branch failed
[info] java.io.IOException: 'GIT_ASKPASS=/home/frank/.github/askpass/scala-steward.sh git -c core.hooksPath=/dev/null checkout test-branch' exited with code 1
```

**Comment by fthomas** [comment]
My last commit fixed the checkout error mentioned above. It then created this PR: https://github.com/scala-steward-org/test-repo-2/pull/49 :tada:

**Comment by alejandrohdezma** [comment]
@fthomas I'm not sure I understand enough the changes you made in #2294 to be able to resolve the conflicts in this PR 😖

**Comment by fthomas** [comment]
I created some merge conflicts with #2294. Sorry about that @alejandrohdezma. That PR was meant to reduce the diff here and I was just about resolving those at the same time as you pushed your latest commits.

**Comment by alejandrohdezma** [comment]
> I created some merge conflicts with #2294. Sorry about that @alejandrohdezma. That PR was meant to reduce the diff here and I was just about resolving those at the same time as you pushed your latest commits.
> 
> Update: I'll resolve them now.

No problemo :)

Should I start with documentation now that the code seems to be ready?

**Comment by fthomas** [comment]
> Maybe `parent.default_branch` just needs to be replaced with `repo.branch.getOrElse(parent.default_branch)`?

I'll revert this since `parent.default_branch` is already the same as `repo.branch`. I didn't realized that `createForkOrGetRepoWithBranch` already modified the parent's `default_branch`.

**Comment by alejandrohdezma** [comment]
I've added [this PR](https://github.com/scala-steward-org/scala-steward-action/pull/290) in the `scala-steward-action` to support this new feature.

**Comment by alejandrohdezma** [comment]
I noticed this feature didn't take into account users running Scala Steward for a GitHub App. For those users the only option would be to have something in the repo (some kind of file) that lists the branches to update and request that file via API.

The only problem is that the REST API has a rate-limit of 5000 calls per-hour, so I investigated it using the GraphQL API (that allows to query several repos at once) and this is the output:

Screenshot 2021-10-17 at 13 50 55

So to sum up, we could change the way the branch is added, so it is added to the `.scala-steward.conf` file (instead of adding `:branch`) to the repos list. And read and parse that file at startup for all repos in the list, reading a new option `branches` from that file.

What do you think @fthomas?

**Comment by alejandrohdezma** [comment]
For example this is the output for the first 100 files listed in the GitHub `repos.md` file:

```json
{
  "data": {
    "r1": {
      "object": null
    },
    "r2": {
      "object": null
    },
    "r3": {
      "object": null
    },
    "r4": {
      "object": null
    },
    "r5": {
      "object": null
    },
    "r6": {
      "object": null
    },
    "r7": {
      "object": {
        "text": "updates.includeScala = true\n"
      }
    },
    "r8": {
      "object": null
    },
    "r9": {
      "object": null
    },
    "r10": {
      "object": {
        "text": "updates.ignore = [\n  { groupId = \"com.typesafe\", artifactId = \"ssl-config-core\" },\n\n  { groupId = \"com.typesafe.akka\", artifactId = \"akka-actor\" },\n  { groupId = \"com.typesafe.akka\", artifactId = \"akka-actor-typed\" },\n  { groupId = \"com.typesafe.akka\", artifactId = \"akka-discovery\" },\n  { groupId = \"com.typesafe.akka\", artifactId = \"akka-slf4j\" },\n  { groupId = \"com.typesafe.akka\", artifactId = \"akka-stream\" },\n  { groupId = \"com.typesafe.akka\", artifactId = \"akka-stream-testkit\" },\n  { groupId = \"com.typesafe.akka\", artifactId = \"akka-testkit\" },\n  { groupId = \"com.typesafe.akka\", artifactId = \"akka-multi-node-testkit\" },\n\n  { groupId = \"com.typesafe.akka\", artifactId = \"akka-http\" },\n  { groupId = \"com.typesafe.akka\", artifactId = \"akka-http-core\" },\n\n  { groupId = \"org.scalatest\", artifactId = \"scalatest\" },\n]\n\nupdates.pin = [\n  { groupId = \"org.scalatest\", artifactId = \"scalatest\", version = \"3.1.\" },\n]\n"
      }
    },
    "r11": {
      "object": null
    },
    "r12": {
      "object": null
    },
    "r13": {
      "object": null
    },
    "r14": {
      "object": null
    },
    "r15": {
      "object": null
    },
    "r16": {
      "object": null
    },
    "r17": {
      "object": null
    },
    "r18": {
      "object": {
        "text": "updates.ignore = [ { groupId = \"org.apache.spark\", artifactId = \"spark-sql\" } ]\n"
      }
    },
    "r19": {
      "object": null
    },
    "r20": {
      "object": null
    },
    "r21": {
      "object": null
    },
    "r22": {
      "object": null
    },
    "r23": {
      "object": {
        "text": "updates.ignore = [ { groupId = \"org.apache.spark\", artifactId = \"spark-sql\" } ]\n"
      }
    },
    "r24": {
      "object": null
    },
    "r25": {
      "object": null
    },
    "r26": {
      "object": null
    },
    "r27": {
      "object": null
    },
    "r28": {
      "object": null
    },
    "r29": {
      "object": null
    },
    "r30": {
      "object": null
    },
    "r31": {
      "object": null
    },
    "r32": {
      "object": null
    },
    "r33": {
      "object": null
    },
    "r34": {
      "object": null
    },
    "r35": {
      "object": null
    },
    "r36": {
      "object": null
    },
    "r37": {
      "object": null
    },
    "r38": {
      "object": null
    },
    "r39": {
      "object": null
    },
    "r40": {
      "object": null
    },
    "r41": {
      "object": null
    },
    "r42": {
      "object": null
    },
    "r43": {
      "object": null
    },
    "r44": {
      "object": {
        "text": "updatePullRequests = \"always\"\ncommits.message = \":arrow_up: update ${artifactName} from ${currentVersion} to ${nextVersion}\"\n"
      }
    },
    "r45": {
      "object": null
    },
    "r46": {
      "object": null
    },
    "r47": {
      "object": null
    },
    "r48": {
      "object": null
    },
    "r49": {
      "object": null
    },
    "r50": {
      "object": null
    },
    "r51": {
      "object": null
    },
    "r52": {
      "object": null
    },
    "r53": {
      "object": null
    },
    "r54": {
      "object": null
    },
    "r55": {
      "object": null
    },
    "r56": {
      "object": null
    },
    "r57": {
      "object": null
    },
    "r58": {
      "object": null
    },
    "r59": {
      "object": null
    },
    "r60": {
      "object": null
    },
    "r61": {
      "object": null
    },
    "r62": {
      "object": null
    },
    "r63": {
      "object": null
    },
    "r64": {
      "object": null
    },
    "r65": {
      "object": null
    },
    "r66": {
      "object": null
    },
    "r67": {
      "object": null
    },
    "r68": {
      "object": null
    },
    "r69": {
      "object": null
    },
    "r70": {
      "object": null
    },
    "r71": {
      "object": null
    },
    "r72": {
      "object": null
    },
    "r73": {
      "object": null
    },
    "r74": {
      "object": null
    },
    "r75": {
      "object": null
    },
    "r76": {
      "object": null
    },
    "r77": {
      "object": null
    },
    "r78": {
      "object": null
    },
    "r79": {
      "object": null
    },
    "r80": {
      "object": null
    },
    "r81": {
      "object": null
    },
    "r82": {
      "object": null
    },
    "r83": {
      "object": null
    },
    "r84": {
      "object": null
    },
    "r85": {
      "object": null
    },
    "r86": {
      "object": null
    },
    "r87": {
      "object": null
    },
    "r88": {
      "object": null
    },
    "r89": {
      "object": null
    },
    "r90": {
      "object": null
    },
    "r91": {
      "object": null
    },
    "r92": {
      "object": {
        "text": "updates.pin  = [ { groupId = \"org.typelevel\", artifactId=\"cats-effect\", version = \"2.\" } ]\n"
      }
    },
    "r93": {
      "object": null
    },
    "r94": {
      "object": null
    },
    "r95": {
      "object": null
    },
    "r96": {
      "object": null
    },
    "r97": {
      "object": null
    },
    "r98": {
      "object": null
    },
    "r99": {
      "object": null
    },
    "r100": {
      "object": null
    }
  }
}
```

**Comment by fthomas** [comment]
> I noticed this feature didn't take into account users running Scala Steward for a GitHub App.

Oh nooo. :scream: That's the curse of having so many features. Can't we just not support branch selection if it used as GitHub App?

**Comment by alejandrohdezma** [comment]
Nothing that I can think of :)

**Comment by fthomas** [comment]
My public instance tracks `master` and does not wait for releases. So every new feature is available instantly.

Should we publish a new version for the GH Action soon?

**Comment by fthomas** [comment]
0.12.0 is now available.
